[MRG] ENH: alternative layout of extra points in interpolation

mmagnuski opened the pull request:
Implements the extra points layout for topo interpolation discussed in #5315 
It may eliminate the issues raised in #5315 but it might be suboptimal in some standard cases. For example this is a frontal theta topography before this PR:
<img src="https://user-images.githubusercontent.com/8452354/49180525-37e00c80-f355-11e8-9013-ee5da9f386ab.png" width="300px">
and this is after:
<img src="https://user-images.githubusercontent.com/8452354/49180535-3f9fb100-f355-11e8-9b53-8aaa26e24a0b.png" width="300px">

I must say I prefer the "before" topo in this case (the "after" looks like a laughing cyclops in a party hat). Probably the extra points are too close to channels. I'll check that later. Also - masking using convex hull (but slightly less extended than here) might make it better.

### TODO:
- [x] rebase and add tests
- [x] add whats new
- [x] link topomap example to evoked plotting tutorial
- [x] enhance example to show the new extrapolation option
- [x] add back the old behavior (let's call it `'box'` bacause the additional points were added on a rectangular box)
- [x] resolve what to do when `outlines='skirt'`
- [x] add tests
- [x] special case for colinear points when `extrapolate='local'`
- [x] test extra points on the head circle
- [x] use median inter-channel distance
- [x] take convex hull from triangulation to use Delaunay only once

larsoner commented:
Personally I don't mind the second one.

We have already discussed having a new param `border=0.` (or `fill_value=0.`) to say what the values should be for these extended edge points. We could add an option (default for backward compat, probably) as `border='extrapolate'`, meaning that these new points get assigend the average value of its neighbors in the Delaunay triangulation of the expanded set of vertices. This would probably make the second one look more like the first.

mmagnuski commented:
I like `boarder=0.` and `boarder='extrapolate'`. But I am still wondering if the added points are'nt too close to the channels. If they were further away the transition would be smoother. Here you can see the scatter outlining their position (you may also notice that the edges of the hull are not divided into equal parts, I can change that in the implementation):
![image](https://user-images.githubusercontent.com/8452354/49181802-a5da0300-f358-11e8-8b8b-4aa12e33bcdd.png)
Also - I am currently calculating the distance as the mean of the 10% lowest distances, which may be too convoluted.

larsoner commented:
> I am still wondering if the added points are'nt too close to the channels.

It makes sense to me to have the edge/"stop interpolating" point be beyond your "average" electrode spacing. It's roughly equivalent to thinking of your data at each electrode representing a value at the center of pixel within grid.

In addition to be a simple way of framing things, it has the advantage of not letting you think that you have data outside that range, which becomes especially important after picking channels. Pick a larger spacing, and your difference between the picked and un-picked representations could very well become worse.

Speaking of which, somewhere we set a mask for the points. Currently it's the head circle. Eventually, we probably want it to be this ConvexHull polygon. This will make the PR much more annoying :) And it will also interact with outlines='skirt' and such... part of the reason it's difficult to make these sorts of changes is we have many of these going on. But I'd first target the standard (non-skirt) full coverage and, say, one-picked-hemisphere data to get them to look reasonable. Then deal with additional cases.

> If they were further away the transition would be smoother.

I suspect that this will mostly be alleviated somewhat by the `extrapolate` option.

> I am currently calculating the distance as the mean of the 10% lowest distances, which may be too convoluted.

We'll have to see how it does in testing. I'd use the median neighbor-to-neighbor (from the Delaunay triangulation) distance because it's the simplest reasonably representative thing I can think of.

mmagnuski commented:
I agree with all your points. 

> I'd use the median neighbor-to-neighbor (from the Delaunay triangulation) distance because it's the simplest reasonably representative thing I can think of.

Yes, the 10% is a temporary hack  :) I will change the distance to the median neighbour-to-neighbor.
From what I remember Delaunay allows to add points after initial computation, so I'll compute it at the beginning for `pos` and later add the additional points to it. If so - would it be better to take the convex hull from Delaunay (although the docs warn that this may be inaccurate) or compute it separately?

> Speaking of which, somewhere we set a mask for the points pcolormesh plot. 

Oh, yes, I am aware of the mask - I was playing with it recently to generate the masked topos I posted in #5315. I can work on that after this PR.
And while you are mentioning skirt - EEGLAB IIRC has a principled way of setting which electrodes are within the head circle, and which outside - channels below some defined y position (similar to 10-20 skull circumference level) are plotted outside the head.

cbrnr commented:
Regarding the added points, I thought they should be exactly on the head circle. Can you try and see how this looks, including setting the values to zero?

mmagnuski commented:
@cbrnr - the added values are zero in the "after" example (that's why the colors go to white around these added points). If they were placed on the head circle as it is drawn:
* the interpolation function would need to know where the head circle is
* some of the channels would have these additional points closer than currently (that is below the average/median inter-electrode distance)
* in the case of plotting only some of the channels and once head circle does not adapt to electrode's range (Eric's PR) these distances might be much higher (for example - plotting occipital channels only would still interpolate to the front of the head)
* then, the interpolation function would also need to know whether the skirt mode is used in which case the zero points shouldn't be added at the circle, but somewhere else
but I can try that - so you can see how it looks, in the case above it should be pretty similar.
@all - If you can paste a few snippets with problematic/interesting topo activity (and channel pos) I could generate more plots for better comparisons.

cbrnr commented:
@mmagnuski but to me it looks like this is what the MATLAB toolbox in https://github.com/mne-tools/mne-python/issues/5315#issuecomment-439803612 is doing - the values on the circle are exactly zero, and this looks like it should.

cbrnr commented:
@jsosulski can you share the code snippet you used to produce topoplots with the MATLAB toolbox?

jsosulski commented:
Sure, see this [gist](https://gist.github.com/jsosulski/c978f81e81be832e52def32eb91a8c01).

Additionally another example how bbci plots topomaps:

![image](https://user-images.githubusercontent.com/2545339/49217837-ff7d1480-f3ce-11e8-86f0-9878cf9fe5bc.png)

In my opinion, the first one is the best compromise between correctness and aesthetically pleasing. The second one does not look good but is 'most accurate' while the last one seems to be misleading at best.

> (the "after" looks like a laughing cyclops in a party hat).

While true, I must say in the first one the outer values look very close to the strong frontal/central activity which could be misinterpreted as activity in that region. This does not happen in the second case.

mmagnuski commented:
I'll check the placement of extra points on the circle later today (the example below show that it may be useful). But using the circle would have to mean that for skirt we still use something like the convex hull. Currently the data provided by @jsosulski give this with this PR (note that the stretched to fill the whole scalp but that is another problem: #5085):
![image](https://user-images.githubusercontent.com/8452354/49223099-08c1ad80-f3de-11e8-9cd3-509917bb0f94.png)

cbrnr commented:
Thanks for the gist @jsosulski - I think MNE also needs an option to not extrapolate at all. As you say, extrapolating is making up information, so we primarily do this for aesthetic purposes. So maybe there could be a parameter `extrapolate` which can be `'head'` (the head circle, default), `'none'` (or `None`), and `'smooth'`. If `extrapolate == 'head'`, we'd also need to specify the values on the head circle, which by default should be `0`. This probably requires another parameter.

I am also +1 for a Pinocchio parameter 😄.

jsosulski commented:
> Currently the data provided by @jsosulski give this with this PR

This looks much better than the current default mne topomap!

> for example - plotting occipital channels only would still interpolate to the front of the head

To compromise, I would use the current method to place these support points, but if they extend beyond the outline of the scalp, put them exactly on the scalp border. This would also (i think) improve the bottom right blue blob in your plot @mmagnuski , which currently does not seem to go towards 0 fast enough.

cbrnr commented:
> At this point my main questions is: what should be the default for extrapolate ('head' vs let's say 'local' or 'hull')? If 'head' then it probably should switch to convex hull in outlines='skirt'. If 'local'/'hull' should be the default then there could be an error on outlines='skirt' + extrapolate='head'. What do you think?

Honestly, I don't understand how these parameters influence the plot. Can't we come up with less technical ones (i.e. use similar parameters to the MATLAB toolbox)? Basically, I want to be able to specify if we extrapolate or not. And if we extrapolate, I want to be able to say to which locations we extrapolate, and what the value at these locations is (default zero). I have difficulties understanding what you mean by skirt, local, and (convex) hull, and probably this will also be problematic for other users.

mmagnuski commented:
> Honestly, I don't understand how these parameters influence the plot. Can't we come up with less technical ones (i.e. use similar parameters to the MATLAB toolbox)?

Sure, these were just my propositions. But to clarify what I meant I respond below:

> I want to be able to specify if we extrapolate or not.

This would come in the next PR. In this one you could specify how you want to extrapolate (see below).

> And if we extrapolate, I want to be able to say to which locations we extrapolate

You could choose whether additional points (the locations to which you want to extrapolate) are placed on the head circle (`extrapolate='head'`) or locally (so it could be `extrapolate='local'`), based on the convex hull (this is just a shape that encompasses all channel positions - similar to the yellow line that you drew in #5315). We could also consider whether one could specify the max distance from the channels to which interpolation is done (this wouldn't be precisely max distance, but the distance by which the convex hull is extended). 

> and what the value at these locations is (default zero).

That would be `boarder=0` if we want to extrapolate to 0 or `boarder='closest'` if the values at the "extrapolation boarder" are based on values of closest channels. 

> I have difficulties understanding what you mean by skirt

Skirt is an option that is currently available in topo plots, it is designed to match what EEGLAB does (channels below the head circumference are moved outside the head circle). See here for an example:
![image](https://user-images.githubusercontent.com/8452354/49308302-31c66900-f4d8-11e8-9151-260037485738.png)

mmagnuski commented:
@larsoner What do you think? And I think we should fall back to the old behavior if there is < 3 channels (this is why CI's fail).

larsoner commented:
> And I think we should fall back to the old behavior if there is < 3 channels

If by "old behavior" you mean "interpolate to the edges" I don't think that's right, at least for the ConvexHull-style interp. In Convex-hull mode, if there are only co-linear points (i.e., <= 2 points, or > 2 but all along a single line) then we can deal with it by putting a rectangle around them (1x the median distance between points away; if only a single point, then just use some sane, small value like 1/20th or 1/10th of the head diameter; in fact maybe the convex hull should always be the `max(median(dists), 0.05)` or something).

cbrnr commented:
Thanks @mmagnuski, this makes sense. We should make/update a topoplot tutorial that shows all options once this and the next PR are done. BTW, the argument should be named `border` and not `boarder`.

mmagnuski commented:
@larsoner Yes - I didn't mean exactly what was previously done but some box-like approach. I'll check the max(median, 0.05) too (although I am not sure at which point the channels are scaled to the head circle).

> We should make/update a topoplot tutorial that shows all options once this and the next PR are done.

Oh, yes, definitelly!

> BTW, the argument should be named border and not boarder.

Yes, I must have meant a cross between `boar` and `border` ;)

mmagnuski commented:
Comparison of `extrapolate='local'` and `extrapolate='head'`, both with border=0:
![image](https://user-images.githubusercontent.com/8452354/49339426-d14b3f00-f631-11e8-9051-bf319edc5df5.png)
![image](https://user-images.githubusercontent.com/8452354/49339442-f2ac2b00-f631-11e8-8904-5b0a7ead0f73.png)
in the case of placing points on head, I actually had to place them slightly outside head (at `0.53` instead of `0.5`) to avoid the pixelated edges.

cbrnr commented:
What does my example from #5315 look like?

mmagnuski commented:
@cbrnr: Both `'local'` and `'head'` look very similar:
![image](https://user-images.githubusercontent.com/8452354/49342691-a4615100-f65e-11e8-82c5-3ffe872b0b2d.png)

mmagnuski commented:
@larsoner 
With respect to `outlines='skirt'` there are (at least) two options:
* make `extrapolate='local'` default and raise error when `extrapolate='head'` and `outlines='skirt'`
* make `extrapolate='head'` default and switch to `extrapolate='local'` when `outlines='skirt'`
I have a preference for the first option as there is less magic under the hood.

cbrnr commented:
Both variants are much better than the previous default, because the red (positive) patches are much less pronounced. However, the [BBCI toolbox topoplot](https://github.com/mne-tools/mne-python/issues/5315#issuecomment-439803612) still looks better because it doesn't contain any positive values.

mmagnuski commented:
@cbrnr - we can check different interpolation methods later to see if that helps. I could also add more points on the border an we can see if that helps.

mmagnuski commented:
@cbrnr Additional points do not help much. Below you can find images of mask highlighting positive values with more (upper plot) and less (lower) extra points. The figures are without `origin='lower'` so you'd have to flip them upside down in your head. :)
![image](https://user-images.githubusercontent.com/8452354/49344135-da5d0000-f673-11e8-9aad-732e3dc6402d.png)

Adding even ludicrous amount of extra points (20 times more than default) doesn't help so much (the plot below is **with** `origin='lower'`, just to confuse you ;) ):
![image](https://user-images.githubusercontent.com/8452354/49344180-2019c880-f674-11e8-99c6-752f2bf3050b.png)

I'd blame it on different interpolator used and maybe some differences in the 2d layout of channels between mne and the BCI toolbox (it seems that 2d channel positions are different - if one could extract the 2d coordinates of the points we could check that).

cbrnr commented:
I suspect its the different interpolation scheme. @jsosulski can you check if there are any positive values in the BBCI topoplot of my example (e.g. with a mask)?

jsosulski commented:
There actually are, I plotted some random integers between 0 and -100 as I did not have your original data available anymore right now. 

![image](https://user-images.githubusercontent.com/2545339/49359364-52f4a880-f6d6-11e8-8cab-2abac9b16ebe.png)

If one uses linear interpolation, there are obviously no more positive values, but the topomap looks more like a picasso picture. For the above plots the matlab toolbox uses griddata with 'v4' interpolation, i.e. [bi-harmonic spline interpolation](https://www.gibboncode.org/html/HELP_biharmonicSplineInterpolation.html).

mmagnuski commented:
With respect to interpolation: it doesn't seem that the bi-harmonic spline interpolation / matlab's V4 is available in scipy. Here is an old discussion at scipy mailing list featuring @agramfort and @dengemann:
https://mail.python.org/pipermail/scipy-dev/2014-July/thread.html#19952
(look for: griddata equivalent of matlab "V4" option)
There is a mention of some PR to scipy, maybe @dengemann knows what happend?

mmagnuski commented:
@larsoner Circle CI is upset with your latest changes: `Invalid requirement: 'dipy --only-binary dipy'` :)

larsoner commented:
If you rebase it should fix it

mmagnuski commented:
@larsoner Also: I am using `np.unique()` with `axis` argument, which was introduced later on in numpy. Should we add a private version of unique with axis to mne?
BTW - I also decided I'd prefer to add the option of interpolating to avg of closest values (not zero) in the follow-up PR, are you ok with that?

larsoner commented:
> Should we add a private version of unique with axis to mne?

Sure, typically we put such backports into `mne/fixes.py` with a copy-paste of the upstream function and note about when they can be removed like
```
# can be removed once we require NumPy 1.YZ
```

> I also decided I'd prefer to add the option of interpolating to avg of closest values (not zero) in the follow-up PR, are you ok with that?

Yes a first PR about boundary choices (this one) and a second one about interpolation value to use at those boundaries sounds cleanest and easiest to review anyway

mmagnuski commented:
@larsoner Great, thanks!

mmagnuski commented on mne/viz/topomap.py:
```diff
@@ -1463,6 +1555,13 @@ def plot_evoked_topomap(evoked, times="auto", ch_type=None, layout=None,
         Defaults to None.
+    extrapolate : str
+        If 'head' extrapolate to the edges of the head circle (does not work
+        when `outlines='skirt'`. If 'local' (default) extrapolate only to
```
I forgot the closing bracket.

mmagnuski commented:
@larsoner @agramfort @cbrnr - I think this is ready for review.
The remaining issues are mostly API-wise:
* I've written a few time here about the interaction between `outlines='skirt'` and `extrapolate='head'`. My preference is to have `extrapolate='local'` as the default and throw an error when `extrapolate='head'` and `outlines='skirt'`.
* I should probably add the extrapolate arg to some more functions that use topo plotting, right? I've added it to evoked's `plot_topomap` but there are surely others like plotting topographies of psd. 

I'll add some more topographical plots in the evening but the only things that I've changed since last plots was fixing cases when points are colinear or there is not enough points for triangulation (< 4 points).

larsoner commented on mne/viz/topomap.py:
```diff
@@ -616,6 +697,13 @@ def plot_topomap(data, pos, vmin=None, vmax=None, cmap=None, sensors=True,
+        If 'head' extrapolate to the edges of the head circle (does not work
+        when `outlines='skirt'`. If 'local' (default) extrapolate only to
+        nearby points (approximately to points closer than median
+        inter-electrode distance).
```
I think there is an API problem here. How do users get the behavior they currently get in `master` for `'skirt'` on your PR? One option would be to keep `'head'` valid for `skirt` mode. Then "head" means "based on the head circle" (which would work and make sense in either `outlines` mode) and `local` means in a small neighborhood around the sensors (wherever they are).

larsoner commented:
> My preference is to have extrapolate='local' as the default and throw an error when extrapolate='head' and outlines='skirt'.

The only problem with changing the default is that it's not backward compatible. How about in this PR you don't change the default, and we can have a follow-up PR to deprecate? It will make reviewing and thinking about problems easier.

> I should probably add the extrapolate arg to some more functions that use topo plotting, right? I've added it to evoked's plot_topomap but there are surely others like plotting topographies of psd.

Yes there are going to be other places it's needed. It can be part of this PR or a separate one, depending on how much time you want to spend on it. When I looked into this code at some point it seemed like there were a lot of difficult interactions and corner cases, so separate PRs might make your life easier.

mmagnuski commented:
+1 to separate PRs, it will make our lifes easier :)
I agree with the backwards compatibility argument - although actually both 'local' and 'head' are not exactly compatible with what was present before (i.e. topographies would be different anyway). So you would prefer to have 'head' as the default - even once your PR that removes scaling channel positions to head circle lands (which should be in 0.18)? It's true that 'head' is more similar to what we had previously but would we want to extrapolate to the whole head circle by default when one uses only a subset of channels?

larsoner commented:
>  It's true that 'head' is more similar to what we had previously but would we want to extrapolate to the whole head circle by default when one uses only a subset of channels?

For now/this PR yes; in the follow-up deprecation PR, no.

My idea is to have this PR only be about adding an `extrapolate` argument whose current default (`'head'`?) does whatever 0.17 does, but adds another option `'local'` that extrapolates to within a small neighborhood of the convex hull of the (flattened) sensor positions.

mmagnuski commented:
skirt with extrapolate='local' does look a bit different from what was before:
![image](https://user-images.githubusercontent.com/8452354/49955393-f9c11c00-ff02-11e8-8924-4726788c3b7b.png)
but the pixelated edges are only temporary - we want to have a convex hull mask anyway, at least for extrapolate='local'.
But I agree it might be better not to raise the error but make extrapolate='head' and outlines='skirt' work peacefully together. Should I set this back to WIP then?

agramfort commented on mne/viz/evoked.py:
```diff
@@ -172,7 +172,8 @@ def _plot_legend(pos, colors, axis, bads, outlines, loc, size=30):
     ax.set_adjustable("box")
-    pos_x, pos_y = _prepare_topomap(pos, ax, check_nonzero=False)
+    pos_x, pos_y = pos.T
+    _prepare_topomap(pos, ax, check_nonzero=False)
```
this is weird. You assign vars and then use them 2 lines later. Please swap the 2 new lines.

agramfort commented on mne/fixes.py:
```diff
@@ -892,6 +892,24 @@ def einsum(*args, **kwargs):
+_has_unique_axis = (LooseVersion(np.__version__) >= '1.13.0')
+
+
+def remove_duplicates(arr):
```
this function should be private and I would put rows in the name

agramfort commented:
these options should be presented in https://martinos.org/mne/dev/auto_tutorials/plot_visualize_evoked.html

mmagnuski commented:
@agramfort Yes, they will be. Actually, @cbrnr suggested to have a separate tutorial on topography plotting. But this PR is just one of multiple steps in that direction, so I think it will be better to add the examples once all the topomap interpolation changes are done.

agramfort commented:
ok but this should be part of this PR. It's important that we consolidate the doc in every PR. If we say later there is a huge risk that it's never done.

mmagnuski commented:
Sure, I understand.

agramfort commented:
Or so we add it to the list of viz tutos and we link to it also from the topomap section of evoked viz ?

mmagnuski commented:
@agramfort I wont have time next week to create a whole new tutorial just for one interpolation option. I can get back to it some time later.

mmagnuski commented:
Ok, another idea: I could modify that example:
https://martinos.org/mne/stable/auto_examples/visualization/plot_evoked_topomap.html

It would require substantially less time and the example looks as if it would benefit from some editing (splitting into code sections for example).

agramfort commented:
Good idea 

Don’t forget to link to it from viz tutorial

mmagnuski commented:
Looks like an improvement.
Before:
https://martinos.org/mne/stable/auto_examples/visualization/plot_evoked_topomap.html

Now:
```
Makefile:49: recipe for target 'html_dev-pattern' failed
make: *** [html_dev-pattern] Error 1
Exited with code 2
```

mmagnuski commented:
a more serious version of now (although I don't know why circle failed, does it fail on warnings now?)
https://10536-1301584-gh.circle-artifacts.com/0/html/auto_examples/visualization/plot_evoked_topomap.html

I like it but would change the size of the huge topo near the end, and would increase the size of the two topos showing extrapolation options.

larsoner commented:
Shouldn't, rather

mmagnuski commented:
Ok, that was my fault, but should be fixed now.

agramfort commented:
can you add sections in your updated example so outline is more visible at
first glance?

mmagnuski commented:
@agramfort I'm not sure if I understand - there are plenty of sections. Do you mean to add a section at the beginning with an overview of the steps?

agramfort commented:
I mean section headers/titles with ---- or ====

larsoner commented:
Yeah it is hard to see the names. I can't think of a way to fix it for now. It's presumably a problem on `master` as well, though so you don't have to fix it here for merge.

mmagnuski commented on mne/viz/topomap.py:
```diff
@@ -616,6 +711,15 @@ def plot_topomap(data, pos, vmin=None, vmax=None, cmap=None, sensors=True,
+        If 'box' (default) extrapolate to four points placed to form a square
+        encompassing all data points, where each side of the square is three
+        times the range of the data in respective dimension. If 'head'
+        extrapolate to the edges of the head circle (does not work when
```
right, although we discussed to add support for skirt + head extrapolation. It might be easy - I'll see what is the mask radius when using skirt.

larsoner replied:
If it does not work immediately or you are feeling lazy, feel free to just raise an error for now and we can add proper support for that case later when someone is motivated

mmagnuski replied:
fixed that (will push soon)

larsoner commented:
If it's huge on master you can just keep it huge for now and we can fix the hugeness later

larsoner commented:
(I don't see a good easy solution so wouldn't want to hold this PR up for it)

mmagnuski commented:
@larsoner I'm using a bad label when linking to the example from evoked viz tutorial. I was just guessing there, how should it be done correctly?

larsoner commented:
`git grep` is your friend here for finding the pattern, but it should be `sphx_glr_auto_examples_visualization_plot_evoked_topomap.py`

larsoner commented:
also don't forget the rebase so that CircleCI can build properly

mmagnuski commented:
Thanks! And: oh, no, another conflict...

mmagnuski commented:
@larsoner @agramfort @cbrnr Travis and circle are green, so it's ready for final round of reviews. Appveyor will take some time until it marches through the queue to the last commit (I would cancel my previous appveyor builds but I do have these privilages on appveyor) - so if one could terminate builds for my previous commits all would go faster. :)

larsoner commented:
Done and +1 for merge

cbrnr commented on mne/viz/topomap.py:
```diff
@@ -470,30 +471,124 @@ def _draw_outlines(ax, outlines):
+        # remove duplicates from hull_extended
+        hull_extended = _remove_duplicate_rows(hull_extended.reshape((-1, 2)))
+        new_pos = np.concatenate([hull_extended] + add_points)
+    else:
```
Maybe add a comment which methods this else branch corresponds to

cbrnr commented on mne/viz/topomap.py:
```diff
@@ -470,30 +471,124 @@ def _draw_outlines(ax, outlines):
+        new_pos = np.concatenate([hull_extended] + add_points)
+    else:
+        # return points on the head circle
+        head_radius = 0.53 if head_radius is None else head_radius
```
Add a short comment to explain the magic number

cbrnr commented on mne/viz/topomap.py:
```diff
@@ -470,30 +471,124 @@ def _draw_outlines(ax, outlines):
+
+    def __init__(self, pos, method='box', head_radius=None):
+        # in principle this works in N dimensions, not just 2
+        assert pos.ndim == 2 and pos.shape[1] == 2
```
I'd replace this with a try/except block

cbrnr commented on mne/viz/topomap.py:
```diff
@@ -616,6 +711,16 @@ def plot_topomap(data, pos, vmin=None, vmax=None, cmap=None, sensors=True,
+    extrapolate : str
+        If 'box' (default) extrapolate to four points placed to form a square
+        encompassing all data points, where each side of the square is three
+        times the range of the data in respective dimension. If 'head'
```
in the respective

cbrnr commented on mne/viz/topomap.py:
```diff
@@ -715,20 +830,14 @@ def _plot_topomap(data, pos, vmin=None, vmax=None, cmap=None, sensors=True,
 
     # interpolate data
+    head_radius = (None if extrapolate == 'local' else
+                   outlines['clip_radius'][0] * 1.06)
```
Explain the magic number

cbrnr commented on mne/viz/topomap.py:
```diff
@@ -1463,6 +1573,16 @@ def plot_evoked_topomap(evoked, times="auto", ch_type=None, layout=None,
+    extrapolate : str
+        If 'box' (default) extrapolate to four points placed to form a square
+        encompassing all data points, where each side of the square is three
+        times the range of the data in respective dimension. If 'head'
```
in the respective

cbrnr commented on mne/viz/topomap.py:
```diff
@@ -2090,14 +2211,15 @@ def _onselect(eclick, erelease, tfr, pos, ch_type, itmin, itmax, ifmin, ifmax,
-    """Prepare the topomap."""
-    pos_x = pos[:, 0]
-    pos_y = pos[:, 1]
+    """
```
Start docstring in same line

cbrnr commented on mne/viz/topomap.py:
```diff
@@ -2537,6 +2660,16 @@ def plot_arrowmap(data, info_from, info_to=None, scale=1e-10, vmin=None,
+    extrapolate : str
+        If 'box' (default) extrapolate to four points placed to form a square
+        encompassing all data points, where each side of the square is three
+        times the range of the data in respective dimension. If 'head'
```
in the respective

cbrnr commented on tutorials/plot_visualize_evoked.py:
```diff
@@ -76,13 +76,14 @@
+# See :ref:`sphx_glr_auto_examples_visualization_plot_evoked_topomap.py` for
+# more advanced topomap plotting options. You can also take a look at the
+# documentation of :func:`mne.Evoked.plot_topomap` or simply write
+# ``evoked_r_aud.plot_topomap?`` in your python console to see the different
```
Python (capital P)

agramfort commented:
+1 for MRG after addressing @cbrnr 's comments

rendered tuto:

https://10563-1301584-gh.circle-artifacts.com/0/html/auto_examples/visualization/plot_evoked_topomap.html#sphx-glr-auto-examples-visualization-plot-evoked-topomap-py

mmagnuski commented on mne/viz/tests/test_topomap.py:
```diff
@@ -157,6 +157,25 @@ def test_plot_topomap():
     evoked.plot_topomap([0.1], ch_type='eeg', scalings=1, res=res,
                         contours=[-100, 0, 100], time_unit='ms')
+
+    # extrapolation to the edges of the convex hull or the head circle
```
No, I don't think so - actually I will be able remove some tests once we change the default (but there will be a deprecation cycle most likely). All the necessary changes will be done in the deprecation PR.

cbrnr replied:
OK, I just wanted to alert you that this test relies on the current default value. If this is what we want, then fine with me.

mmagnuski commented on examples/visualization/plot_evoked_topomap.py:
```diff
@@ -21,30 +24,81 @@
+###############################################################################
+# Compare the image above with the one below, where we use
+# ``extrapolate='head'`` so that extrapolation would go to 0 at the head
+# circle (another option is to use ``extrapolate='local'`` in which case the
```
I didn't want to complicate the plotting with operations on axes, but I can change that.

mmagnuski commented on examples/visualization/plot_evoked_topomap.py:
```diff
@@ -21,30 +24,81 @@
+###############################################################################
+# If you look at the edges of the head circle of a single topomap you'll see
+# the effect of extrapolation. By default ``extrapolate='box'`` is used which
+# extrapolates to large box stretching beyond the head circle:
```
yes, I saw that but was too lazy to correct, thanks :)

mmagnuski commented:
Oh, another conflict, third rebase, hooray. :)

mmagnuski commented:
Thanks for the comments @cbrnr, I think I adressed all. Now I am at the mercy of CIs...

cbrnr commented:
There are still some open comments about magic numbers and a missing "the" in two or more places. Since the CIs have already failed you might as well push these changes.

larsoner commented:
Travis
```
tutorials/plot_visualize_evoked.py:104:80: E501 line too long (80 > 79 characters)
```
CircleCI
```
/home/circleci/project/examples/visualization/plot_evoked_topomap.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/circleci/miniconda/envs/mne/lib/python3.6/site-packages/sphinx_gallery/gen_rst.py", line 402, in _memory_usage
    multiprocess=True)
  File "/home/circleci/miniconda/envs/mne/lib/python3.6/site-packages/memory_profiler.py", line 336, in memory_usage
    returned = f(*args, **kw)
  File "/home/circleci/miniconda/envs/mne/lib/python3.6/site-packages/sphinx_gallery/gen_rst.py", line 393, in __call__
    exec(self.code, self.globals)
  File "/home/circleci/project/examples/visualization/plot_evoked_topomap.py", line 85, in <module>
    show=False)
  File "/home/circleci/project/mne/evoked.py", line 359, in plot_topomap
    axes=axes, extrapolate=extrapolate)
  File "/home/circleci/project/mne/viz/topomap.py", line 1773, in plot_evoked_topomap
    cax = plt.subplot(1, n_fig_axes + 1, n_fig_axes + 1)
  File "/home/circleci/miniconda/envs/mne/lib/python3.6/site-packages/matplotlib/pyplot.py", line 1084, in subplot
    a = fig.add_subplot(*args, **kwargs)
  File "/home/circleci/miniconda/envs/mne/lib/python3.6/site-packages/matplotlib/figure.py", line 1352, in add_subplot
    ax = self._axstack.get(key)
  File "/home/circleci/miniconda/envs/mne/lib/python3.6/site-packages/matplotlib/figure.py", line 98, in get
    "Adding an axes using the same arguments as a previous axes "
  File "/home/circleci/miniconda/envs/mne/lib/python3.6/site-packages/matplotlib/cbook/deprecation.py", line 112, in warn_deprecated
    warnings.warn(message, category, stacklevel=2)
matplotlib.cbook.deprecation.MatplotlibDeprecationWarning: 
Adding an axes using the same arguments as a previous axes currently reuses the earlier instance.  In a future version, a new instance will always be created and returned.  Meanwhile, this warning can be suppressed, and the future behavior ensured, by passing a unique label to each axes instance.
```

mmagnuski commented:
@cbrnr, thanks, I'll fix it later today.

mmagnuski commented:
thanks @larsoner I'm on it.

mmagnuski commented:
https://10598-1301584-gh.circle-artifacts.com/0/dev/auto_examples/visualization/plot_evoked_topomap.html

cbrnr commented:
Thanks @mmagnuski!

mmagnuski commented:
Unbelivable - it's finally merged :)
Thanks @cbrnr @larsoner @agramfort!
:rocket:
